File: 1matlab/SeamCarving.m
function [resizedImage] = ModifiedSeamCarving(ImgName, hor, ver)
InputImage = imread(ImgName);
GrayImage = rgb2gray(InputImage);
GradMean = CalcGradient(GrayImage);
ReducedInputImage = InputImage;
ReducedGrayImage = GrayImage;
ReducedGradMean = GradMean;
for i=1:1:min(ver, hor)
[ReducedInputImage,ReducedGradMean, ReducedGrayImage] = RemoveVer(ReducedGradMean, ReducedInputImage, ReducedGrayImage);
[ReducedInputImage,ReducedGradMean, ReducedGrayImage] = RemoveHor(ReducedGradMean, ReducedInputImage, ReducedGrayImage);
end
for j=1:1:(max(ver,hor)-min(ver,hor))
    if(ver>hor)
        [ReducedInputImage,ReducedGradMean, ReducedGrayImage] = RemoveVer(ReducedGradMean, ReducedInputImage, ReducedGrayImage);
    else
        [ReducedInputImage,ReducedGradMean, ReducedGrayImage] = RemoveHor(ReducedGradMean, ReducedInputImage, ReducedGrayImage);
    end
end
ReducedGrayImage = uint8(ReducedGrayImage);
ReducedInputImage = uint8(ReducedInputImage);
figure
imshow(InputImage);
figure
imshow(ReducedGrayImage);
figure
imshow(ReducedInputImage);
end

% calculating gradient for input image
function [GradMean] = CalcGradient(inputImage)
    sobelOpx = [ 1 0 -1 ; 2  0  -2; 1 0 -1];
    sobelOpy = [1 2 1; 0 0 0; -1 -2 -1];
    [rows, cols, dim] = size(inputImage);
    for i= 1:1:dim
        GradHor(:,:,i) = filter2(sobelOpx, inputImage(:,:,i));
        GradVer(:,:,i) = filter2(sobelOpy, inputImage(:,:,i));
        Grad(:,:,i) = abs(GradHor(:,:,i)) + abs(GradVer(:,:,i));
    end
    GradMean = sum(Grad, 3)/dim;
end

%Remove Vertial Seam
function [ReducedInputImage,ReducedGradMean, ReducedGrayImage] = RemoveVer(ReducedGradMean, ReducedInputImage, ReducedGrayImage)
EnergyMapVer = CalcEnergyMapForward(ReducedGradMean, ReducedGrayImage);
SeamVer = findSeam(EnergyMapVer);
ReducedInputImage = RemoveSeam(ReducedInputImage,SeamVer);
ReducedGrayImage = RemoveSeam(ReducedGrayImage,SeamVer);
ReducedGradMean = RemoveSeamGradient(ReducedGradMean, SeamVer);
end

% Remove Horizontal Seam. It uses same functions as vertical so we need to
% transpose images before use those functions.
function [ReducedInputImage,ReducedGradMean, ReducedGrayImage] = RemoveHor(ReducedGradMean, ReducedInputImage, ReducedGrayImage)
EnergyMapHor = CalcEnergyMapForward(ReducedGradMean', permute(ReducedGrayImage,[2,1,3]));
SeamHor = findSeam(EnergyMapHor);
ReducedInputImage = RemoveSeam(permute(ReducedInputImage, [2,1,3]), SeamHor);
ReducedGrayImage = RemoveSeam(permute(ReducedGrayImage, [2,1,3]), SeamHor);
ReducedGradMean = RemoveSeamGradient(permute(ReducedGradMean,[2,1,3]), SeamHor);
ReducedInputImage = permute(ReducedInputImage, [2,1,3]);
ReducedGradMean = permute(ReducedGradMean, [2,1,3]);
ReducedGrayImage = permute(ReducedGrayImage, [2,1,3]);
end

%Calculating Forward energy map
function [EnergyMap] = CalcEnergyMapForward(GradMean,ReducedGrayImage)
    [rows, cols] = size(GradMean);
  for i = 1:1:rows
    for j = 1:1:cols
      if(i== 1)
        EnergyMap(1,:) = GradMean(1,:);
      else if(j == 1)
        EnergyMap(i,j) = GradMean(i,j)+ min((EnergyMap(i-1,j) +
          costVertical(i,j,ReducedGrayImage)), (EnergyMap(i-1,j+1)+
          costRight(i,j,ReducedGrayImage)));
      elseif(j == cols)
        EnergyMap(i,j) = GradMean(i,j)+ min((EnergyMap(i-1,j-1)+
          costLeft(i,j,ReducedGrayImage)), (EnergyMap(i-1,j)+
          costVertical(i,j,ReducedGrayImage)));
      else
        EnergyMap(i,j) = GradMean(i,j)+ min(min((EnergyMap(i-1,j-1)+
          costLeft(i,j,ReducedGrayImage)), (EnergyMap(i-1,j)+
          costVertical(i,j,ReducedGrayImage))), (EnergyMap(i-1,j+1)+
          costRight(i,j,ReducedGrayImage)));
      end
    end
  end
end

%Additional cost functions for forward energy map
function [value] = costLeft(i,j,ReducedGrayImage)
  [cols] = size(ReducedGrayImage,2);
  if(j==cols)
    value = abs(ReducedGrayImage(i-1,j)-ReducedGrayImage(i,j-1));
  else
    value = abs(ReducedGrayImage(i,j+1)-ReducedGrayImage(i,j-1))+abs(ReducedGrayImage(i-1,j)-ReducedGrayImage(i,j-1));
  end
end

function [value] = costRight(i,j,ReducedGrayImage)
  if(j==1)
    value = abs(ReducedGrayImage(i-1,j)-ReducedGrayImage(i,j+1));
  else
    value = abs(ReducedGrayImage(i,j+1)-ReducedGrayImage(i,j-1))+abs(ReducedGrayImage(i-1,j)-ReducedGrayImage(i,j+1));
  end
end

function [value] = costVertical(i,j,ReducedGrayImage)
  [cols] = size(ReducedGrayImage,2);
  if(j==1 || j==cols)
    value = 0;
  else
    value = abs(ReducedGrayImage(i,j+1)-ReducedGrayImage(i,j-1));
  end
end

%Finding Seam from Energy Map
function[Seam] = findSeam(EnergyMap)
[rows, cols] = size(EnergyMap);
Seam = zeros(rows,1);
for k = 1:1:rows
    if(k==1)
        [Value, Seam(rows)] = min(EnergyMap(rows,:)); 
    else
        if(Seam(rows-k+2) == 1)
             [Value, temp] = min(EnergyMap(rows-k+1,1:2));
             Seam(rows-k+1)= temp;
        elseif(Seam(rows-k+2) == cols)
             [Value, temp] = min(EnergyMap(rows-k+1,cols-1:cols));
             Seam(rows-k+1)= Seam(rows-k+2)-2 + temp;
        else
        [Value, temp] = min(EnergyMap(rows-k+1,Seam(rows-k+2)-1:1:Seam(rows-k+2)+1));
        Seam(rows-k+1)= Seam(rows-k+2)-2 + temp;
        end
    end
end
end

%Removing seam from image
function [ReducedImage] = RemoveSeam(inputImage, Seam)
[rows, cols, dim] = size(inputImage);
ReducedImage = zeros(rows,cols-1, dim);
%imshow(inputImage);
for k=1:1:dim
    for i=1:1:rows
        for j=1:1:cols-1
            if(j < Seam(i))
                ReducedImage(i,j,k) = inputImage(i,j,k);
            elseif(j >= Seam(i))
                ReducedImage(i,j,k) = inputImage(i,j+1,k);
            end                
        end
    end
        
end
end

%Remove Seam from gradient image. We add importance diffusion for gradient
%image in this funciton so we calculated it seperately.
function [ReducedGradMean] = RemoveSeamGradient(GradMean, Seam)
[rows, cols, dim] = size(GradMean);
ReducedGradMean = zeros(rows,cols-1, dim);
for i=1:1:rows
    for j=1:1:cols-1
        %%% Importance Diffusion Starts
%         if(j == Seam(i)-1)
%             GradMean(i,j) = GradMean(i,j) + GradMean(i,Seam(i))/8;
%         end
%         if(j == Seam(i))
%             GradMean(i,j+1) = GradMean(i,j+1) + GradMean(i,Seam(i))/8;
%         end
        %%Importance Diffusion Ends
        if(j < Seam(i))
            ReducedGradMean(i,j) = GradMean(i,j);
        elseif(j >= Seam(i))
            ReducedGradMean(i,j) = GradMean(i,j+1);
        end                
    end
end
end

File: old/single/SeamCarving.cpp
#include "opencv2/imgproc/imgproc.hpp"
#include <iostream>
#include <cstdlib>
#include <cmath>
#include <string>
#include <opencv/cv.h>
#include <opencv/highgui.h>

//include "SeamCarving.h"

using namespace cv;
using namespace std;

void printHelp()
{
  cout << "Usage: SeamCarving <image> <horizontal lines to cut> " <<
    "<vertical lines to cut>" << endl;
}

int costLeft(int i, int j, Mat rGrayImg)
{
  int cols = rGrayImg.rows;
  int value;

  if(j == cols)
    value = (int)(abs(rGrayImg.at<unsigned char>(i-1, j) -
      rGrayImg.at<unsigned char>(i, j-1)));
  else
    value = (int)(abs(rGrayImg.at<unsigned char>(i, j+1) -
      rGrayImg.at<unsigned char>(i, j-1)) + 
      abs(rGrayImg.at<unsigned char>(i-1, j) - 
      rGrayImg.at<unsigned char>(i, j-1)));

  return value;
}

int costRight(int i, int j, Mat rGrayImg)
{
  //int cols = rGrayImg.rows;
  int value;

  if(j == 1)
    value = (int)(abs(rGrayImg.at<unsigned char>(i-1, j) -
      rGrayImg.at<unsigned char>(i, j+1)));
  else
    value = (int)(abs(rGrayImg.at<unsigned char>(i, j+1) -
      rGrayImg.at<unsigned char>(i, j-1)) +
      abs(rGrayImg.at<unsigned char>(i-1, j) - 
      rGrayImg.at<unsigned char>(i, j+1)));

  return value;
}

int costVertical(int i, int j, Mat rGrayImg)
{
  int cols = rGrayImg.rows;
  int value;

  if((j == 1) || (j == cols))
    value = 0;
  else
    value = (int)(abs(rGrayImg.at<unsigned char>(i, j+1) -
     rGrayImg.at<unsigned char>(i, j-1)));

  return value;
}

void calcEnergyMapForward(Mat &rGrad, Mat &rGrayImg,
  Mat &eMapVer)
{
  //int channels = rGrad.channels();
  int nRows = rGrad.rows;
  int nCols = rGrad.cols; // removed: "* channels;", do we need this for 3 color channels?
  int cval1, cval2, cval3;

  // this probably isn't necessary since we are creating this mat as multidim
  if(rGrad.isContinuous())
  {
    nCols += nRows;
    nRows = 1;
  }

  for(int i = 0; i < nRows; ++i)
  {
    for(int j = 0; j < nCols; ++i)
    {
      if(i == 1)
      {
        eMapVer.at<unsigned char>(i, j) = rGrad.at<unsigned char>(i, j);
      }
      else if(j == 1)
      {
        cval1 = eMapVer.at<unsigned char>(i-1, j) + 
          costVertical(i, j, rGrayImg);
        cval2 = eMapVer.at<unsigned char>(i-1, j+1) +
          costRight(i, j, rGrayImg);

        eMapVer.at<unsigned char>(i, j) = rGrad.at<unsigned char>(i, j) +
          ((cval1 < cval2) ? cval1 : cval2);
      }
      else if(j == nCols-1)
      {
        cval1 = eMapVer.at<unsigned char>(i-1, j-1) +
          costLeft(i, j, rGrayImg);
        cval2 = eMapVer.at<unsigned char>(i-1, j) +
          costVertical(i, j, rGrayImg);

        eMapVer.at<unsigned char>(i, j) = rGrad.at<unsigned char>(i, j) +
          ((cval1 < cval2) ? cval1 : cval2);
      }
      else
      {
        int minVal;
        cval1 = eMapVer.at<unsigned char>(i-1, j-1) +
          costLeft(i, j, rGrayImg);
        cval2 = eMapVer.at<unsigned char>(i-1, j) +
          costVertical(i, j, rGrayImg);
        cval3 = eMapVer.at<unsigned char>(i-1, j+1) +
          costRight(i, j, rGrayImg);
        if((cval1 < cval2) && (cval1 < cval3))
          minVal = cval1;
        else if((cval2 < cval1) && (cval2 < cval3))
          minVal = cval2;
        else
          minVal = cval3;

        eMapVer.at<unsigned char>(i, j) = rGrad.at<unsigned char>(i, j) +
          minVal;
      }
    }
  }

}

void findSeam(Mat energyMap, int seam[])
{
    int rows = energyMap.rows;
    int cols = energyMap.cols;
    int *newSeam;
    int minIdx;

    newSeam = new int[rows];

    for(int k = 0; k < rows; ++k)
    {

        if(k == 1)
        {
            minMaxIdx(energyMap.col(rows-1), /*&minVal*/ NULL, NULL, &minIdx, NULL);
            newSeam[rows-1] = minIdx;
        }
        else
        {
            if(newSeam[rows-k+2] == 1)
            {
                minMaxIdx(energyMap.col(rows-k+1).rowRange(Range(0, 1)), NULL, NULL, &minIdx, NULL);
                newSeam[rows-1] = minIdx;
            }
            else if(newSeam[rows-k+2] == cols)
            {
                minMaxIdx(energyMap.col(rows-k+1).rowRange(Range(cols-2, cols-1)), NULL, NULL, &minIdx, NULL);
                newSeam[rows-k+1] = seam[rows-k+2] - 2 + minIdx;
            }
            else
            {
                minMaxIdx(energyMap.col(rows-k+1).rowRange(Range(seam[rows-k+2]-1, seam[rows-k+2])), NULL, NULL, &minIdx, NULL);
                newSeam[rows-k+1] = seam[rows-k+2] -2 + minIdx;
            }
        }
    }

    seam = newSeam;
}

Mat* removeSeam(Mat inputImage, int seam[])
{
    int rows = inputImage.rows;
    int cols = inputImage.cols;
    //Mat reducedImage(rows, cols-1, CV_8UC3);
    Mat *reducedImage = new Mat(Mat::zeros(rows, cols-1, CV_8UC3));

    for(int i = 0; i < rows; ++i)
    {
        inputImage.row(i).colRange(Range(0, seam[i])).copyTo(reducedImage->row(i).colRange(Range(0, seam[i])));
        inputImage.row(i).colRange(Range(seam[i]+1, cols)).copyTo(reducedImage->row(i).colRange(Range(seam[i], cols-1)));
    }

    return reducedImage;
}

Mat* removeSeamGradient(Mat gradMean, int seam[])
{
    int rows = gradMean.rows;
    int cols = gradMean.cols;
    //Mat reducedGradMean(cols-1, rows, CV_8UC3);
    Mat *reducedGradMean = new Mat(Mat::zeros(rows, cols-1, CV_8UC3));

    for(int i = 0; i < rows; ++i)
    {
        gradMean.row(i).colRange(Range(0, seam[i])).copyTo(reducedGradMean->row(i).colRange(Range(0, seam[i])));
        gradMean.row(i).colRange(Range(seam[i]+1, cols)).copyTo(reducedGradMean->row(i).colRange(Range(seam[i], cols-1)));
    }

    return reducedGradMean;
}

void removeVer(Mat &reducedInputImage, Mat &reducedGrayImage,
  Mat &reducedGradMean)
{
  Mat energyMapVer;
  int *seamVer;

  calcEnergyMapForward(reducedGradMean, reducedGrayImage, energyMapVer);
  findSeam(energyMapVer, seamVer);
  removeSeam(reducedInputImage, seamVer);
  removeSeam(reducedGrayImage, seamVer);
  removeSeamGradient(reducedGradMean, seamVer);
}

void removeHor(Mat &reducedInputImage, Mat &reducedGrayImage,
  Mat &reducedGradMean)
{
  Mat energyMapHor;
  int *seamHor;

  reducedGradMean.t();
  reducedGrayImage.t();
  calcEnergyMapForward(reducedGradMean, reducedGrayImage,
    energyMapHor);
  findSeam(energyMapHor, seamHor);
  reducedInputImage.t();
  reducedGrayImage.t();
  removeSeam(reducedInputImage, seamHor);
  removeSeam(reducedGrayImage, seamHor);
  removeSeamGradient(reducedGradMean, seamHor);
}

int main(int argc, char* argv[])
{
  Mat inputImage, grayImage, gradMean;
  Mat reducedInputImage, reducedGrayImage, reducedGradMean;
  char* filename;
  int hor, ver;
  int diffHorVer, minDim;

  if(argc < 4)
  {
    printHelp();
    return -1;
  }

  filename = argv[1];
  hor = atoi(argv[2]);
  ver = atoi(argv[3]);

  inputImage = imread(filename, 1);

  if(!inputImage.data)
  {
    cout << "Unable to open image file." << endl;
    return -1;
  }

  // create greyscale and edge-detected images
  cvtColor(inputImage, grayImage, CV_RGB2GRAY);
  Sobel(grayImage, gradMean, 1, 0, 0, 3, 1, 0, BORDER_DEFAULT);

  // once this is all working, get rid of these reduced images and use the
  // original ones
  inputImage.copyTo(reducedInputImage);
  grayImage.copyTo(reducedGrayImage);
  gradMean.copyTo(reducedGradMean);

  // calculate the difference between the 2 number of cuts and the min of them
  if(hor > ver)
  {
    diffHorVer = hor - ver;
    minDim = ver;
  }
  else
  {
    diffHorVer = ver - hor;
    minDim = hor;
  }

  // do alternating horizontal and vertical seam carves up the the min of
  // the two
  for(int i = 0; i < minDim; ++i)
  {
    removeVer(reducedInputImage, reducedGrayImage, reducedGradMean);
    removeHor(reducedInputImage, reducedGrayImage, reducedGradMean);
  }

  // finish carving the larger dimension
  if(hor > ver)
    for(int i = 0; i < diffHorVer; ++i)
    {
      removeHor(reducedInputImage, reducedGrayImage, reducedGradMean);
    }
  else
    for(int i = 0; i < diffHorVer; ++i)
    {
      removeVer(reducedInputImage, reducedGrayImage, reducedGradMean);
    }

  // create output filename and save resized image
  string sFilename = filename;
  string sFront = sFilename.substr(0, sFilename.find(".", 0));
  string sEnd = sFilename.substr(sFilename.find(".", 0), string::npos);
  string sRenamed = sFront + "-carved" + sEnd;
  imwrite(sRenamed.c_str(), reducedInputImage);

  return 0;
}
